includes/content.php

```diff
@@ -179,96 +191,139 @@ function pmpro_search_filter($query)
+	 * Note this also gets IDs for posts and other custom post types
+	 * that might be protected through the PMPro CPT Add On or
+	 * other plugins or custom code.
+	 */
+	// First, figure out if their is a current user with levels.
+	if( ! empty( $current_user->ID ) ) {
+		$levels = pmpro_getMembershipLevelsForUser($current_user->ID);
+	} else {
+		$levels = false;
+	}
+	if( ! empty( $levels ) ) {
+		$level_ids = wp_list_pluck( $levels, 'ID' );
+	}
+
+	// Now get posts hidden by level.
+	if( ! empty( $level_ids ) ) {
+		// All hidden pages, minus the ones I have access to.
+		$sql = "SELECT mp.page_id
+				FROM {$wpdb->pmpro_memberships_pages} mp
+					LEFT JOIN {$wpdb->posts} p ON mp.page_id = p.ID
+				WHERE mp.page_id NOT IN(
+					SELECT page_id
+					FROM {$wpdb->pmpro_memberships_pages}
+					WHERE membership_id IN(" . implode(',', array_map( 'esc_sql', $level_ids ) ) . ")
+				) AND p.post_type IN('" . implode( "', '", array_map( 'esc_sql', $pmpro_search_filter_post_types ) ) . "')";
+	} else {
+		// All hidden pages.
+		$sql = "SELECT mp.page_id
+				FROM {$wpdb->pmpro_memberships_pages} mp
+					LEFT JOIN {$wpdb->posts} p ON mp.page_id = p.ID
+				WHERE p.post_type IN('" . implode( "', '", array_map( 'esc_sql', $pmpro_search_filter_post_types ) ) . "')";
+	}
+	$hidden_post_ids = array_values(array_unique($wpdb->get_col($sql)));
+	
+	/**
+	 * Get post IDs hidden by category.
+	 * Note this also gets IDs for tags and other custom taxonomies
+	 * since the PMPro table doesn't differentiate between them.
+	 */
+	if ( ! empty( $level_ids ) ) {
+		// All posts hidden by category, minus the ones I have access to.
+		$sql = "SELECT tr.object_id
+				FROM {$wpdb->term_relationships} tr
+					LEFT JOIN {$wpdb->posts} p ON tr.object_id = p.ID
+				WHERE tr.term_taxonomy_id IN(					
+					SELECT category_id
+					FROM {$wpdb->pmpro_memberships_categories}
+					WHERE membership_id NOT IN(" . implode(',', array_map( 'esc_sql', $level_ids ) ) . ")
+				) AND tr.object_id NOT IN(					
+					SELECT object_id
+					FROM {$wpdb->term_relationships}
+					WHERE term_taxonomy_id IN(
+						SELECT category_id
+						FROM {$wpdb->pmpro_memberships_categories}
+						WHERE membership_id IN(" . implode(',', array_map( 'esc_sql', $level_ids ) ) . ")
+					)
+				) AND p.post_type IN(
+					'" . implode( "', '", array_map( 'esc_sql', $pmpro_search_filter_post_types ) ) . "'
+				)";
+	} else {
+		// All posts hidden by category.
+		$sql = "SELECT tr.object_id
+				FROM {$wpdb->term_relationships} tr
+					LEFT JOIN {$wpdb->posts} p ON tr.object_id = p.ID
+				WHERE tr.term_taxonomy_id IN(
+					SELECT category_id
+					FROM {$wpdb->pmpro_memberships_categories}
+				) AND p.post_type IN('" . implode( "', '", array_map( 'esc_sql', $pmpro_search_filter_post_types ) ) . "')";
+	}
+	$cat_hidden_post_ids = array_values(array_unique($wpdb->get_col($sql)));
+
+	// Merge the hidden post IDs.
+	$hidden_post_ids = array_merge( $hidden_post_ids, $cat_hidden_post_ids );
```

Comment:
Do we want to `array_unique()` this or does it not matter?

Reply:
I think we should use array_unique here. It might not matter, but feels cleaner and helps with any other array stuff that might be done later by others filtering the query.